### ISSUE TITLE
Gossipsub parameter tuning

### ISSUE THREAD

**Issue by AgeManning** [ethereum/consensus-specs#2692]
# Overview

The gossipsub mesh parameters were originally chosen for message propagation speed and redundancy (a large number of faulty nodes shouldn't impact the message propagation too much). However, now that our networks have advanced, become stable and have a significant number of nodes, it might be time to re-visit the gossipsub parameters. 

We (Lighthouse) have been looking into bandwidth consumption of our nodes and are in the process of doing some more analysis (which we'll share when we have it) and ways to reduce this without degrading our current networks.

The primary source of excessive bandwidth consumption is our gossipsub network which has a large degree of amplification on all our topics. 

## Potentially Low-Hanging Fruit

I've been monitoring the number of duplicates we have been receiving on various topics as well as the number of IWANT messages we have been sending. Both of these are good measures of redundant bandwidth. The goal, is to minimise these for each topic, without degrading the network. 

A node typically sends IWANT messages when they receive an IHAVE from one of their peers for a message before they receive it from one of their mesh peers. This suggests the mesh is propagating slower or on the order of a node's peer's heartbeat interval.  Messages can be delayed on the mesh for a number of reasons; message validation, network size, mesh connectivity, etc. Each client has their own speeds for validation, but also peer counts. A client that maintains more peer's is more likely to receive IHAVE's from one of their peers before their mesh (as the mesh size is bounded, regardless of peer count). 

One suggestion to reducing IWANT/IHAVE is to either increase our gossipsub `heartbeat_interval` and/or decrease our `D_lazy` parameter to reduce the amount of gossiping each node does. These changes can however have an impact of message propagation latency, if nodes are currently relying on the gossip IWANT/IHAVE system to get their messages. 

## Message Amplification

We have set `D` to be `8` in the specs which naturally gives us relatively high message amplification. On the block topic (the simplest to reason about), we see on average about 8 duplicates (as would naively be expected) and therefore a bandwidth amplification of about 8x. The other topics are less straightforward. 

We can significantly reduce our node's bandwidth by reducing the mesh sizes. The impact of this at a network-level is non-trivial. As we flood-publish, a client's default peer count plays a large role in how many hops or how fast a message propagates as we publish to every peer we know who is connected on that topic. Therefore, it may be possible to have higher peer counts, and lower mesh sizes whilst maintaining a similar number of hops to reach the entire network.

In any case our mesh sizes appear to be larger than needed and we may want to lower them (incrementally on testnets). 

## Proposal

I think it would be useful to put some lower bounds on the gossipsub mesh topic parameters. For lighthouse, we are considering having some of the parameters CLI tuneable. i.e users with no bandwidth limitation can have a high degree of connectivity (which provides them with lower message latency) and have some users with lower bandwidth but higher message latency.  We were going to leave the default as is, so that most users will have no change to the current gossipsub settings.

However to do this safely, there needs to be some lower bound on the mesh sizes such that if all nodes choose the lowest setting, the network still has sufficient redundancy/amplification and message propagation speed.

Ultimately, I think we should try test some parameters on testnets (slowly) before moving towards mainnet. The parameters can be lowered slowly and iterated on over time. 

I know that a full simulation/analysis of the live network would be ideal before changing/justifying these parameters, however I don't know of anyone that can/has modelled our current eth2 network and can definitively say how much effect adjusting these parameters across all nodes on the network will have. Our testnet's also don't have the same topology as mainnet, so tests there may not reflect on mainnet. 

With that caveat aside, I'm suggesting we add some lower bounds to the gossipsub values, something like:

- `D` : 5 (maybe even 4 as a lower bound)
- `D_low`: 3
-  `D_high`: 6
- `D_lazy`: 3 (or maybe 2)
- `heartbeat_interval` : Maybe up to half a slot (clients can choose when they want to send gossip) 

Note that these would be lower bounds and clients can maintain their current values and no work needs to be done. However we discuss and allow for some implementations to reduce these values.

We may also want to modify our implementations such that we can specify these on a per-topic basis, if more analysis is done and better values can be found per-topic.

**Comment by arnetheduck** [comment]
> Some topics naturally have much more duplicates 

How is message volume related to duplication?

**Comment by nisdas** [comment]
> How is message volume related to duplication?

Duplication might be the wrong word, but more `IHAVE` messages relaying the same message ids.

**Comment by jmozah** [comment]
>The proposed change would be to not only store the message-id from peers, but also the peers that have sent us this exact message. Then when we propagate/publish, we check that a peer hasn't already sent us this exact message. If a peer has sent us this exact message, its a waste for us to send it to them.

Another navie way could be to append the last few overlays prefix (first 2 bytes maybe) that the message has travelled and ignore retransmitting to similar overlays.  This method can still have duplicates, but theoretically contained under a small neighbourhood of the entire network which have similar overlay prefixes (how small is based on the prefix length we chose to use).

ofcourse this has a disadvantage that peers will know to some extent from which direction the message is coming from.

**Comment by arnetheduck** [comment]
> I think this is relevant even without message validation

How so? Presumably, without validation you receive a message and rebroadcast it to everyone without delay - ie there's no "gap" where nodes can send you messages and thus signal that they have it already - unless one is introduced artificially.

> overhead of storing

Just thought a little bit about this, and the overhead should actually be negligible - we only need to store this information for messages that are in-validation - basically, the first time we see the message and send it off to validation, we register it for monitoring - then, when others send it as well, we record that, then when validation is done, we send it to anyone that wasn't in the monitoring record and remove the information completely. Subsequent messages will be discarded by the "seen" cache regardless and won't ever enter the validation pipeline so there's no overhead for them.

> counter that bumps

This is an interesting idea, but it also introduces state that must be maintained across across the two clients, which I think adds a complication that we should strive to avoid, exhausting other possibilities first.

**Comment by AgeManning** [comment]
> How so? Presumably, without validation you receive a message and rebroadcast it to everyone without delay - ie there's no "gap" where nodes can send you messages and thus signal that they have it already - unless one is introduced artificially.

Yep you're right. Its only in the validation gap this needs to be dealt with. Your suggestion is good. I've modified my implementation as to how we are storing the peers. Shifted it to the mem-cache and its a bit cleaner and should reduce significantly the memory footprint.

**Comment by leolara** [comment]
I am closing this issue because it seems stale. Please, do not hesitate to reopen it if this is a mistake
